sys/kern/kern_cheri_revoke.c

```diff
@@ -315,6 +315,12 @@ kern_cheri_revoke(struct thread *td, int flags,
 	vm = td->td_proc->p_vmspace;
 	vmm = &vm->vm_map;
 
+	if ((flags & (CHERI_REVOKE_LAST_PASS | CHERI_REVOKE_ASYNC)) ==
+	    (CHERI_REVOKE_LAST_PASS | CHERI_REVOKE_ASYNC)) {
+		/* Not sure what this means yet, so don't allow it. */
+		return (EINVAL);
+	}
```

Comment:
Hm.  Since the load-based revoker really just has the one pass, `LAST_PASS` is probably not a good name any more.  Cornucopia could, we thought, meaningfully do N not-`LAST_PASS` passes and then just the one, though that turned out not to be useful (the knobs you rightfully deleted earlier in this series were intended to allow concurrent userspace some finer cooperative control of a sequence of passes, but that, too, turned out to be mostly moot)...

It's not even really clear that it's worth having not-`LAST_PASS` `cheri_revoke()`s be a thing.  I think I/we used to think that giving userspace policy control of whether starting revocation implied wanting to finish it as soon as possible would be a useful thing, but... I no longer remember _why_ I thought that, and I don't think I was right in any case. :)

Maybe cutting userspace's control down to "I would like epoch N to be in the past, and I am {not,} willing to block until that is true" makes sense?

Reply:
Yes, LAST_PASS is a bit confusing. :) For now I'm content to have the two mutually exclusive flags which effectively just determine whether the revocation scan will be synchronous or not, but some further simplification is definitely possible.